File: C++/Medium/Accepted/possible-bipartition_7.cpp
class Solution {
public:
    bool possibleBipartition(int n, vector<vector<int>>& dislikes) {
        
        vector<vector<int>> g(n+1);
        vector<int> color(n+1, 0);
        for (auto& p : dislikes) {
            g[p[0]].push_back(p[1]);
            g[p[1]].push_back(p[0]);
        }
        for (int i = 1; i <= n; i++) {
            if (color[i] == 0 && !dfs(g, color, 1, i)) {
                return false;
            }
        }
        return true;
    }
    bool dfs(vector<vector<int>>& g, vector<int>& color, int c, int i) {
        if (color[i] != 0) {
            return color[i] == c;
        }
        color[i] = c;
        for (auto& j : g[i]) {
            if (!dfs(g, color, -c, j)) {
                return false;
            }
        }
        return true;
    }
};

File: Rust/Medium/Accepted/possible-bipartition_7.rs
impl Solution {
    pub fn possible_bipartition(n: i32, dislikes: Vec<Vec<i32>>) -> bool {
        
        let mut edges = vec![vec![]; n as usize];
        for dislike in dislikes {
            let (a, b) = (dislike[0] as usize - 1, dislike[1] as usize - 1);
            edges[a].push(b);
            edges[b].push(a);
        }
        let mut colors = vec![0; n as usize];
        for i in 0..n as usize {
            if colors[i] == 0 {
                let mut queue = std::collections::VecDeque::new();
                queue.push_back(i);
                colors[i] = 1;
                while !queue.is_empty() {
                    let node = queue.pop_front().unwrap();
                    for &edge in edges[node].iter() {
                        if colors[edge] == colors[node] {
                            return false;
                        }
                        if colors[edge] == 0 {
                            colors[edge] = -colors[node];
                            queue.push_back(edge);
                        }
                    }
                }
            }
        }
        true
    }
}